pxtblocks/fields/field_melodySandbox.ts

```diff
@@ -0,0 +1,407 @@
+            value = value.trim(); // remove boundary white space
+            let melodies: string[] = value.split("-");
+            this.createMelodyIfDoesntExist();
+            this.setTitle(melodies[0]);
+            this.setTempo(Number(melodies[1]));
+            for (let i = 2; i < melodies.length - 1; i++) { // first two strings are name and tempo
+                let notes: string[] = melodies[i].split(" ");
+                for (let j = 0; j < notes.length - 1; j++) {
+                    if (notes[j] != "R") {
+                        let rowPos: number = pxtmelody.noteToRow(notes[j]);
+                        this.melody.updateMelody(rowPos, j);
+                    }
+                }
+            }
+        }
+
+        // The width of the preview on the block itself
+        protected getPreviewWidth(): number {
+            this.updateWidth();
+            return this.size_.width;
+        }
+
+        // The height of the preview on the block itself
+        protected getPreviewHeight(): number {
+            return Blockly.BlockSvg.FIELD_HEIGHT;
+        }
+
+        protected getDropdownBackgroundColour() {
+            return this.sourceBlock_.getColour();
+        }
+
+        protected getDropdownBorderColour() {
+            return this.sourceBlock_.getColourSecondary();
+        }
+
+        setTitle(melodyName: string): void {
+            if (this.title != melodyName) {
+                this.title = melodyName;
+                Blockly.FieldLabel.prototype.setText.call(this, this.title);
+                if (this.melody) {
+                    this.melody.setTitle(this.title);
+                }
+            }
+        }
+
+        getTitle(): string {
+            return this.title;
+        }
+
+        setTempo(tempo: number): void {
+            // reset text input if input is invalid
+            if (isNaN(tempo) || tempo <= 0 && this.tempoInput) {
+                this.tempoInput.value = this.tempo + "";
+                return
+            }
+            // update tempo and duration values and display to reflect new tempo
+            if (this.tempo != tempo) {
+                this.tempo = tempo;
+                if (this.melody) {
+                    this.melody.setTempo(this.tempo);
+                }
+                if (this.tempoInput) {
+                    this.tempoInput.value = this.tempo + "";
+                }
+            }
+        }
+
+        // ms to hold note
+        getDuration(): number {
+            return 60000 / this.tempo;
+        }
+
+        createMelodyIfDoesntExist(): boolean {
+            if (!this.melody) {
+                this.melody = new pxtmelody.MelodyArray();
+                return true;
+            }
+            return false;
+        }
+
+        onNoteSelect(id: string): void {
+            // parse element id
+            let params = id.split("-"); // params[1] is row, params[2] is cell
+            let row = params[1];
```

Comment:
Just convert `row` and `col` to numbers here instead of doing it in place below.